scylla/src/statement/prepared_statement.rs

```diff
@@ -95,34 +97,36 @@ impl PreparedStatement {
             }
             return Ok(buf.into());
         }
-        // TODO: consider what happens if a prepared statement is of type (?, something, ?),
-        // where all three parameters form a partition key. The middle one is not available
-        // in bound values.
-
-        // TODO: Optimize - maybe we could check if pk_indexes are sorted and do an allocation-free two-pointer sweep algorithm then?
-        // We can't just sort them because the hash will break:
-        // https://github.com/apache/cassandra/blob/caeecf6456b87886a79f47a2954788e6c856697c/doc/native_protocol_v4.spec#L673
-
-        let values: Vec<Option<&[u8]>> = bound_values.iter().collect();
+        let mut pk_values: SmallVec<[_; 8]> = smallvec![None; self.metadata.pk_indexes.len()];
+        let mut values_iter = bound_values.iter();
+        let mut offset = 0;
+        let mut buf_size = 0;
         for pk_index in &self.metadata.pk_indexes {
             // Find value matching current pk_index
-            let next_val: &Option<&[u8]> = values
-                .get(*pk_index as usize)
-                .ok_or_else(|| PartitionKeyError::NoPkIndexValue(*pk_index, bound_values.len()))?;
-
-            // Add value's bytes
+            let next_val = values_iter
+                .nth((pk_index.index - offset) as usize)
```

Comment:
Just to clarify, could you explain why `offset` is needed here? Is it for a case where somebody sends a (legal) prepared statement with weird order, like `(partition_key_part1, partition_key_part2, regular_value, partition_key_part3)`, and we need to skip `regular_value` when iterating over bound values? Or is there any other reason for its existence? Note that it would still not solve the TODO mentioned in my previous comment, because we still don't have full information in bound_values to reconstruct the partition key.

Reply:
The `offset` is not a special case, it's needed as part of the algorithm.
Because `pk_indexes` contains indexes, current offset in the values iterator needs to be kept in order to know how many values it should parse to get the next partition key one. For example, if indexes are 1 and 2, you have to call `nth(1)` and then `nth(0)`; at the start, offset is 0, so `1-0 = 1`, after the first `nth`, offset is 2, so `2 - 2 = 0`. 
Actually, `offset` is maybe not the best name, `values_iter_offset` should be more explicit. I will add a comment to explain the algorithm.